### ISSUE TITLE
Unable to define scalar hints for abstract classes

### ISSUE THREAD

**Issue by shadowhand** [rdlowrey/auryn#133]
```php
class Dep
{
}

abstract class Foo
{
    public function __construct($key, Dep $dep)
    {
        print_r(compact('key', 'dep'));
    }
}

class Bar extends Foo
{
}

$injector = new Auryn\Injector;
$injector->define('Foo', [
    ':key' => 'secret',
]);
print_r($injector->make('Bar'));exit;
```

This currently fails with:

```
Auryn\InjectionException: No definition available to provision typeless parameter $key at position 0 in Foo::__construct()
```

It seems like #30 provided partial support for abstracts, but not when it comes to scalar parameters? The current workaround is:

```php
$injector->defineParam('key', 'secret');
```

Which works globally but is not ideal.

**Comment by kelunik** [comment]
This has nothing to do with scalar type hints. It's because you defined `:key` for `Foo` instead of `Bar`.

**Comment by bvisness** [comment]
@kelunik The problem is if you have multiple classes like `Bar` that extend `Foo`, and you want them to all have the same value for `key`.

**Comment by kelunik** [comment]
Yes, but it has nothing to do with scalar type declarations. It's because Auryn doesn't take the definitions for parent classes into account.

**Comment by shadowhand** [comment]
@kelunik I'll rename the PR.

**Comment by morrisonlevi** [comment]
Auryn intentionally does not make children from definitions of their parents. As mentioned in chat the other day this tends to come up often, so we should make a FAQ section and add it to it.

**Comment by morrisonlevi** [comment]
The FAQ should also provide alternatives. Sometimes it is appropriate to `alias`; sometimes not. What other choices are there?

**Comment by kelunik** [comment]
Indeed, the child constructor can be entirely different and things should not be inherited IMO. We should add something to a FAQ.
